src/main/sensors/esc_sensor.c

```diff
@@ -1518,6 +1440,658 @@ static void apdSensorProcess(timeUs_t currentTimeUs)
+        len += (tribParamAddrLen[j] & TRIB_PAL_LEN_MASK);
+    return len;
+}
+
+static bool tribBuildNextParamReq(void)
+{
+    // pending reset request...
+    if (tribResetEsc) {
+        tribResetEsc = false;
+
+        tribUncSetup = TRIB_UNCSETUP_INACTIVE;
+        tribInvalidParams = 0;
+        tribDirtyParams = 0;
+        paramMspActive = false;
+
+        uint16_t reset = 0;
+        tribBuildReq(0xD0, 0x00, &reset, 2, TRIB_PARAM_FRAME_PERIOD, TRIB_PARAM_WRITE_TIMEOUT);
+        return true;
+    }
+
+    // ...or pending write request...
+    uint8_t offset = 0;
+    for (uint16_t *pal = tribParamAddrLen, dirtybits = tribDirtyParams; dirtybits != 0; pal++, dirtybits >>= 1) {
+        const uint8_t addr = (*pal & TRIB_PAL_ADDR_MASK) >> 8;
+        const uint8_t len = *pal & TRIB_PAL_LEN_MASK;
+        if ((dirtybits & 0x01) != 0) {
+            void *payload;
+            uint32_t iq22Payload[2];
+            if (addr == TRIB_PARAM_IQ22_ADDR) {
+                // convert scaled uint32 -> IQ22
+                const uint32_t q22 = 1 << 22;
+                uint32_t *pw = iq22Payload;
+                uint32_t *pp = (uint32_t*)(paramUpdPayload + offset);
+                *pw++ = (((double)*pp++) / 100 * q22);
+                *pw = (((double)*pp) / 100000 * q22);
+                payload = iq22Payload;
+            }
+            else {
+                payload = paramUpdPayload + offset;
+            }
+            tribBuildReq(TRIB_PKT_WRITE_SETTINGS, addr, payload, len, TRIB_PARAM_FRAME_PERIOD, TRIB_PARAM_WRITE_TIMEOUT);
+            return true;
+        }
+        offset += len;
+    }
+
+    // ...or schedule pending read request if no pending writes...
+    for (uint16_t *pal = tribParamAddrLen, invalidbits = tribInvalidParams; invalidbits != 0; pal++, invalidbits >>= 1) {
+        if ((invalidbits & 0x01) != 0) {
+            uint8_t req = (*pal & TRIB_PAL_SYS_MASK) != 0 ? TRIB_PKT_READ_SYSTEM : TRIB_PKT_READ_SETTINGS;
+            uint8_t addr = (*pal & TRIB_PAL_ADDR_MASK) >> 8;
+            uint8_t len = *pal & TRIB_PAL_LEN_MASK;
+            tribBuildReq(req, addr, NULL, len, TRIB_PARAM_FRAME_PERIOD, TRIB_PARAM_READ_TIMEOUT);
+            return true;
+        }
+    }
+
+    // ...or nothing pending
+    return false;
+}
+
+static bool tribDecodeReadParamResp(uint8_t addr)
+{
+    uint8_t offset = 0;
+    for (uint8_t i = 0; i < ARRAYLEN(tribParamAddrLen); i++) {
+        const uint16_t *pal = tribParamAddrLen + i;
+        const uint8_t len = *pal & TRIB_PAL_LEN_MASK;
+        if ((*pal >> 8) == addr) {
+            // cache params by addr
+            if (addr == TRIB_PARAM_IQ22_ADDR) {
+                // convert IQ22 -> scaled uint32
+                const uint32_t q22 = 1 << 22;
+                uint32_t *pr = (uint32_t*)(buffer + TRIB_HEADER_LENGTH);
+                uint32_t *pp = (uint32_t*)(paramPayload + offset);
+                *pp++ = round(((double)*pr++) / q22 * 100);
```

Comment:
Same question about the use of double.

Reply:
As above. 
These are the gov P and I parameters - one option would be not to present them but some like to use the ESC's internal governor and expect to be able to tweak these. Odd choice by Scorpion.

Reply:
float replaces double